### ISSUE TITLE
(Self hosted) Cloud Storage issue

### ISSUE THREAD

**Issue by marcb199** [webrcade/webrcade#70]
Running version 0.0.5 in Docker. Everything works except linking to Dropbox. Getting this error message: "This WebRcade Server is not authorized to access Dropbox."
Is there a fix for this? 
Thanks in advance.

**Comment by raz0red** [comment]
Unfortunately, Dropbox is not compatible with local hosting of webrcade (Docker, etc.). 

In the future I plan on adding a local Dropbox equivalent that runs on the Docker image itself.

**Comment by raz0red** [comment]
Another option that would be simpler in the short term would be to provide instructions for users to configure their own "webrcade app" in Dropbox.

**Comment by marcb199** [comment]
Hi, would be really great if you could provide those instructions. Thanks!

**Comment by raz0red** [comment]
It will require some tweaks, but shouldn’t be too bad. I will try to put something together in the next couple of days.

**Comment by marcb199** [comment]
Really appreciate it, thank you!

**Comment by KingIzzymon** [comment]
@marcb199 do you still need help with this issue?

**Comment by marcb199** [comment]
Hi, yes! Any help setting this up would be great.

**Comment by KingIzzymon** [comment]
# Install with Docker

The fastest way to get up and running is via the [Docker container](https://hub.docker.com/r/webrcade/webrcade).
After you have docker installed on either [Linux](https://docs.docker.com/engine/install/ubuntu/) or [Windows](https://docs.docker.com/desktop/install/windows-install/), do one of the follow:

Install via docker cli:

```sh
docker run -rm \
  --name=webrcade \
  -p 80:80 \
  -v /path/to/content:/var/www/html/content \
  --restart unless-stopped \
  webrcade/webrcade:latest
```

or deploy with docker-compose:

```yaml
version: "3"
services:
  webrcade:
    image: webrcade/webrcade:latest
    container_name: webrcade
    volumes:
      - /path/to/content:/var/www/html/content # Docker for Windows use "/c/path/to/content"
    ports:
      - 80:80
    restart: unless-stopped
```

## Setup

The `/path/to/content` location will be accessible to the webRcade server so it can host your games and feeds locally.

Its recommended to have your ROMs in their own system folder structure:

### Example File Structure
    - nes/Super Mario Bros. (U).nes (or .zip)
    - gba/Sonic Advance (U).gba (or .zip)
    - myfeed.json

When setting up games/feeds with the editor, use the following example URLs:

### Example URLs
    - /content/nes/Super Mario Bros. (U).nes
    - /content/gba/Sonic Advance (U).gba
    - /content/myfeed.json

**Comment by KingIzzymon** [comment]
@marcb199 If this works out for you, please close the ticket.

- Always remember that, being a product in early development, things may change in the future. Checkout the [release notes](https://github.com/webrcade/webrcade/releases) for any new changes.

**Comment by andrewmackrodt** [comment]
I discovered WebЯcade recently and started adding it to my self-hosted stack today and stumbled across this issue. Having a quick look at the code, the Dropbox account settings appear to be hardcoded in [webrcade/webrcade-app-common](https://github.com/webrcade/webrcade-app-common/blob/master/src/storage/cloud/dropbox/index.js). I'm not sure of the best way to make this configurable (as a PR) as it gets built into several bundles. One approach may be to:

- add a `config.json` file which defaults to an empty object (`{}`) in the dist bundle and can be overridden by docker users
- read the config into `window.globals.config` when the webpage is requested
- modify the implementation of `storage/cloud/dropbox/index.js` to use values from `window.globals.config` and default to the static properties already defined in the file if they do not exist

For now, to test that this may be all that is required, I've modified the start-up command of my container and can confirm that saving and loading states works. This requires creating a new [Scoped App (App Folder)](https://www.dropbox.com/developers/apps/) in Dropbox and assigning it the correct scopes. I'm not sure what scopes are required but have selected: `files.metadata.write`, `files.metadata.read`, `files.content.write`, `files.content.read`. I've also set the OAuth 2 Redirect URIs but do not know if this step is required:

- https://play.mydomain.com/dropbox/
- https://play.mydomain.com/app/editor/dropbox/
- https://play.mydomain.com/app/standalone/dropbox/

I'm using traefik as a reverse proxy and coincidentally was already using subpath rewriting so that accessing my host by `play.mydomain.com` displays the `/play` subpath. However, the redirect back from Dropbox expects `/play` in the url path so I've added an additional rule to redirect to the non `/play` path.

My service definition is as below (note that I'm using an M1 MacBook, defining `platform` and modifying `/etc/apache2/apache2.conf` shouldn't be needed on `linux/amd64` platforms which is what I imagine most people will be using):

```yaml
---
services:
  webrcade:
    image: webrcade/webrcade
    platform: linux/amd64
    labels:
      - traefik.enable=true
      - traefik.http.middlewares.webrcade-play-redirect.redirectRegex.regex=^(https?://[^/]+)/play\b(.*)$$
      - traefik.http.middlewares.webrcade-play-redirect.redirectRegex.replacement=$${1}$${2}
      - traefik.http.middlewares.webrcade-play-prefix.addPrefix.prefix=/play
      - traefik.http.routers.webrcade.middlewares=webrcade-play-redirect,webrcade-play-prefix
      - traefik.http.routers.webrcade.rule=HostRegexp(`play.mydomain.com`)
      - traefik.http.services.webrcade.loadBalancer.server.port=80
    environment:
      - HTTP_HOST=play.mydomain.com
      - WEBRCADE_DROPBOX_APP_KEY=changeme
    entrypoint: /bin/sh
    command:
      - -c
      - |-
        find /var/www/html -type f -name '*.js' -or -type f -name '*.js.map' \
          | xargs grep -l 'play.webrcade.com' \
          | xargs sed -i "s/play.webrcade.com/$$HTTP_HOST/g"
        find /var/www/html -type f -name '*.js' -or -type f -name '*.js.map' \
          | xargs grep -l 'xvz6ybh9wta2u30' \
          | xargs sed -i "s/xvz6ybh9wta2u30/$$WEBRCADE_DROPBOX_CLIENT_ID/g"
        if ! grep -qE '^Mutex' /etc/apache2/apache2.conf; then
          echo "Mutex posixsem" >>/etc/apache2/apache2.conf
        fi
        exec /usr/local/bin/docker-php-entrypoint /home/start.sh "$@"
    volumes:
      - ./config/webrcade/content:/var/www/html/content
    restart: unless-stopped
```

In summary, a hack/workaround is to replace the hardcoded redirect uris and app key on container start.

**Comment by raz0red** [comment]
Great, thanks a lot for the feedback! I am actively working to get a new build out this upcoming week that includes some changes to Docker (specifically an ARM64 image). I don't think I will be able to integrate your changes this round, but I can prioritize for the next update.

I never really anticipated anyone else using Dropbox for saves for local deployments, but I can add the customization hooks similar to how you have described to allow for it. 

Thanks again for the contribution!

**Comment by nopper** [comment]
Hitting the same problem but I'd like to know if there's a better way to solve this for self-hosted scenario. Maybe adding the possibility of storing saved state directly on the server (e.g. upload to the server, webdav, minio, or another storage api)? I guess this is connected to your comment:

> Unfortunately, Dropbox is not compatible with local hosting of webrcade (Docker, etc.).
> 
> In the future I plan on adding a local Dropbox equivalent that runs on the Docker image itself.

**Comment by raz0red** [comment]
Yeah, in the future, we will add the ability to either use Docker as the host (but will require a bit of work on your end). Or, just integrate a completely different storage option into the Docker image (this would be my preference).

**Comment by bphenriques** [comment]
Picking up on this. I am interested in self-hosting and added some of my findings in [here](https://github.com/webrcade/webrcade/issues/7#issuecomment-1874111696). This seems to be right place to discuss.

In short:
- Ability to load read-only files from local volumes.
- Automatically generate a feed from a local directory.
- Work with a non-root user. I am somewhat "new" to permissions and docker, so I am not entirely sure. I can tell we might need to create a separate volume to store logs where the user can write into.
- Ability to use a reverse proxy (i.e. define a `BASE_URL`).

The pattern I am trying to use:
```yaml
version: "3.8"

services:
  # https://github.com/webrcade/webrcade/issues/7#issuecomment-1314400463. Exposes port 80.
  webrcade:
    image: webrcade/webrcade:latest
    container_name: webrcade
    hostname: webrcade
    # user: "${PUID}:${PGID}" # Does not work as it cannot write to logs!
    volumes:
      - "./config:/var/www/html/content"
      - "${GAMING_ROMS_DIR}/gba:/var/www/html/content/gba:ro" # read-only intentionlly 
    restart: unless-stopped
    networks:
      - webrcade

networks:
  webrcade:
    name: webrcade
```
